Run html linter as github action

FiskFan1999 opened the pull request:
Hey everyone. I configured a workflow to run an HTML linter on each commit and pull request in this repository. This way, if there are any syntax errors it will show up as a failed test (red X next to the commit), and this will help test html during pull requests as well.

The workflow syntax is not too complicated to understand, I can walk you through it if need be. This can also be configured to use a different linter, or multiple at the same time.

The linter used is [htmlhint, part of megalinter](https://megalinter.github.io/5.8.0/descriptors/html_htmlhint/)

You can see this workflow being used on the main branch of https://github.com/FiskFan1999/obiesource.github.io (you can submit a PR to my fork if you like, to see how the test fails or passes based on the linter)

Aurelius333 reviewed (approved):
This is fine with me, except a few of the comments (lines 1, 5, and 15) seem to be out of date.

Also, maybe we should use [actions/setup-node@v2](https://github.com/actions/setup-node) instead of `sudo apt install npm -y`. Using `apt` gives this message: `WARNING: apt does not have a stable CLI interface. Use with caution in scripts.`

Note that since we don't give HTMLHint any configuration, by default, it seems to use these rules: <https://github.com/htmlhint/HTMLHint/blob/4ad3fe424e064b03d338852af2997b93332ef31d/src/core/core.ts#L14-L23> (I couldn't find this in the documentation). All other rules are disabled.

FiskFan1999 commented:
> This is fine with me, except a few of the comments (lines 1, 5, and 15) seem to be out of date.
> 
> Also, maybe we should use [actions/setup-node@v2](https://github.com/actions/setup-node) instead of `sudo apt install npm -y`. Using `apt` gives this message: `WARNING: apt does not have a stable CLI interface. Use with caution in scripts.`
> 
> Note that since we don't give HTMLHint any configuration, by default, it seems to use these rules: https://github.com/htmlhint/HTMLHint/blob/4ad3fe424e064b03d338852af2997b93332ef31d/src/core/core.ts#L14-L23 (I couldn't find this in the documentation). All other rules are disabled.

Thank you for your suggestions. I switched to [actions/setup-node@v2](https://github.com/actions/setup-node) and that is working nicely. Will be more optimized and protect against backward-compatibility breaks as well.

As for htmlhint: I found [these rules](https://gist.github.com/FiskFan1999/86a9ab2724cf41fac3677454bcc6beb3) and selected all of the defaults some more that I thought were appropriate (I think some of them are mutually exclusive). These extra rules resulted in a few errors which is the source of the other file changes. I also added the lint itself in the Makefile so those who have npm installed can run it locally.

AjaiKN commented:
Great!

I would propose removing the `tags-check` rule. I can't find any reason that it's good to add the `title` attribute to `<a>`s. It's just duplicate information. Apparently the `title` attribute [usually isn't recommended for accessibility](https://www.a11yproject.com/posts/title-attributes/).

AjaiKN commented on .github/workflows/htmlhint_linter.yml:
```diff
@@ -0,0 +1,21 @@
+      
+    - name: Linter
+      # https://github.com/htmlhint/HTMLHint
+      run: npx htmlhint -r "tagname-lowercase,attr-lowercase,attr-value-double-quotes,doctype-first,doctype-html5,tag-pair,spec-char-escape,id-unique,src-not-empty,attr-no-duplication,title-require,attr-unsafe-chars,href-abs-or-rel,html-lang-require,id-unique,input-requires-label,tag-pair,tagname-lowercase,tagname-specialchars,tags-check,attr-no-unnecessary-whitespace,empty-tag-not-self-closed"
```
I don't think we should have duplication between the list of rules in the GitHub Action and the Makefile. 2 possible solutions:

1. Run `make lint` instead of `npx ...` in the GitHub Action
2. [Create a config file](https://htmlhint.com/docs/user-guide/configuration)

FiskFan1999 commented:
commited 4cdcd93142a51c32566685579b8c7cb7e0f86b7f for comments by @Aurelius333

AjaiKN commented on index.html:
```diff
@@ -7,6 +7,6 @@
     <div class="main">
         <h1>ObieSource</h1>
-        <a href="/membersdir.html">Browse our members directory!</a>
+        <a title="Directory of club members" href="/membersdir.html">Browse our members directory!</a>
```
I would also suggest removing the `title` attributes that you added. (here, in `member.html`, and in `membersdir.html`)

wknowleskellett commented on .github/workflows/htmlhint_linter.yml:
```diff
@@ -0,0 +1,21 @@
+    - uses: actions/checkout@v2
+    - uses: actions/setup-node@v2
+      
+    - name: htmlLimt
```
^ typo

wknowleskellett commented on .github/workflows/htmlhint_linter.yml:
```diff
@@ -0,0 +1,21 @@
+
+    steps:
+    - uses: actions/checkout@v2
+    - uses: actions/setup-node@v2
```
per [this page](https://docs.github.com/en/actions/learn-github-actions/understanding-github-actions#create-an-example-workflow) these can be `@v3`. I don't know what the difference is, is there a reason to use `@v2`?

AjaiKN replied:
It looks like some of the docs were only updated to v3 a few days ago, so that's probably the only reason we're using v2.

wknowleskellett commented on Makefile:
```diff
@@ -1,2 +1,5 @@
 	python3 -m http.server --bind 127.0.0.1 8000
+
+lint: 
+	npx htmlhint -r "tagname-lowercase,attr-lowercase,attr-value-double-quotes,doctype-first,doctype-html5,tag-pair,spec-char-escape,id-unique,src-not-empty,attr-no-duplication,title-require,attr-unsafe-chars,href-abs-or-rel,html-lang-require,id-unique,input-requires-label,tag-pair,tagname-lowercase,tagname-specialchars,attr-no-unnecessary-whitespace,empty-tag-not-self-closed"
```
If we make an `.htmlhintrc`, this line can be `npx htmlhint`, and each of these comma-separated items can be on their own line, like [here](https://gist.github.com/oogatta/3654604f9dff9347beb76900a95b6da5). Since git tracks changes line-by-line, this means that changes to the configuration identify the specific lines that were changed.
I look to make it a config file the moment it's more than just a few items long

FiskFan1999 replied:
Thank you @wknowleskellett , i'll take care of these when i resubmit

wknowleskellett reviewed (changes requested):
I'm fine with the @v2 mentioned and will accept that with or without justification, but the typo should be fixed, and I'd prefer to see a config file. Somebody else can accept the changes without a config file though and I wouldn't be upset.
